add content-filtered-topic interfaces

iuhilnehc-ynos opened the pull request:
Related to https://github.com/ros2/design/pull/282 to add content-filtered-topic interfaces.

fujitatomoya commented on rcl/include/rcl/subscription.h:
```diff
@@ -204,6 +205,104 @@ RCL_WARN_UNUSED
+rcl_ret_t
+rcl_subscription_options_fini(rcl_subscription_options_t * option);
+
+/// Check if the content filter topic feature is supported in the subscription.
```
```suggestion
/// Check if the content filtered topic feature is supported in the subscription.
```

iuhilnehc-ynos replied:
updated

fujitatomoya commented on rcl/include/rcl/subscription.h:
```diff
@@ -204,6 +205,104 @@ RCL_WARN_UNUSED
+ * This function will set a filter expression and an array of expression parameters
+ * for the given subscription, but not to update the original rcl_subscription_options_t
+ * of subscription.
+ *
```
I think it would be probably better to explain the example how to use especially `filter_expression` and `expression_parameters` and how to reset (or clear) the filtering on the subscription.

iuhilnehc-ynos replied:
I tried to update more detailed information by adding some DDS specifications about CFT.

fujitatomoya commented on rcl/include/rcl/subscription.h:
```diff
@@ -204,6 +205,104 @@ RCL_WARN_UNUSED
+rcl_ret_t
+rcl_subscription_get_cft_expression_parameters(
+  const rcl_subscription_t * subscription,
+  char ** filter_expression,
```
do we expect multiple filtering expression here? according to [DDS 1.4](https://www.omg.org/spec/DDS/1.4/PDF) 2.2.2.3.3 ContentFilteredTopic, it is single string? (same question goes to `rmw_subscription_get_cft_expression_parameters`)

iuhilnehc-ynos commented:
Close this PR because #901 includes these updates.

fujitatomoya commented on rcl/src/rcl/subscription.c:
```diff
@@ -193,6 +210,181 @@ rcl_subscription_get_default_options()
+  rcl_subscription_content_filter_options_t * options
+)
+{
+  (void)subscription;
+  (void)options;
+
+  return RCL_RET_UNSUPPORTED;
```
why these are not supported? is it supposed to call `rmw_subscription_get_content_filter`?

fujitatomoya commented on rcl/src/rcl/subscription.c:
```diff
@@ -225,6 +417,7 @@ rcl_take(
   if (!taken) {
     return RCL_RET_SUBSCRIPTION_TAKE_FAILED;
   }
+
```
can we delete this unnecessary change?

fujitatomoya commented:
@ivanpauno could you review this?

iuhilnehc-ynos commented on rcl/include/rcl/context.h:
```diff
@@ -43,6 +43,9 @@ extern "C"
 /// A unique ID per context instance.
 typedef uint64_t rcl_context_instance_id_t;
 
+/// Common content filter factory (Defined in a new package)
+typedef struct common_content_filter_factory_s common_content_filter_factory_t;
+
```
Yes

iuhilnehc-ynos commented on rcl/include/rcl/context.h:
```diff
@@ -314,6 +317,27 @@ RCL_WARN_UNUSED
+ * \param[in] context object from which the common content filter factory should be retrieved.
+ * \return pointer to common content filter factory if valid, otherwise `NULL`
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+common_content_filter_factory_t *
+rcl_context_get_common_content_filter_factory(rcl_context_t * context);
```
Yes, I will remove this interface. Using `#include "./context_impl.h"` and to directly access the `impl->common_content_filter_factory` is enough.

iuhilnehc-ynos commented on rcl/src/rcl/subscription.c:
```diff
@@ -193,6 +210,181 @@ rcl_subscription_get_default_options()
+  const rcl_subscription_content_filter_options_t * options
+)
+{
+  (void)subscription;
+  (void)options;
+
+  return RCL_RET_UNSUPPORTED;
```
I think that we just want to expose all the interfaces (notice that the `rcl_subscription_is_cft_enabled` is removed) in this PR before April 4, 2022.

fujitatomoya replied:
The plan is to support CFT interfaces for rcl and rclcpp, so that user application can take advantage of CFT for user detined messages.

iuhilnehc-ynos commented on rcl/src/rcl/subscription.c:
```diff
@@ -193,6 +210,181 @@ rcl_subscription_get_default_options()
+rcl_ret_t
+rcl_subscription_options_set_content_filter_options(
+  const char * filter_expression,
+  size_t expression_parameters_argc,
```
I intend to check this limitation in `rcl_subscription_set_content_filter` and only be checked inside `rcl_subscription_set_content_filter` once.

fujitatomoya replied:
that makes sense to me.

iuhilnehc-ynos commented on rcl/include/rcl/subscription.h:
```diff
@@ -209,6 +218,165 @@ RCL_WARN_UNUSED
+rcl_ret_t
+rcl_subscription_options_fini(rcl_subscription_options_t * option);
+
+/// Set the content filter options for the given subscription options.
```
What about updating the message at https://github.com/ros2/rcl/blob/0ca51058da545522d13c4fb304be3cae53969a29/rcl/include/rcl/subscription.h#L319

https://github.com/ros2/rcl/blob/0ca51058da545522d13c4fb304be3cae53969a29/rcl/include/rcl/subscription.h#L333 is used to set content filter after subscription is created, and https://github.com/ros2/rcl/blob/0ca51058da545522d13c4fb304be3cae53969a29/rcl/include/rcl/subscription.h#L265 is used while creating a subscription.

fujitatomoya commented:
@Barry-Xu-2018 @ivanpauno requesting another review. I will also follow this up in the weekend.

ivanpauno commented on rcl/src/rcl/subscription.c:
```diff
@@ -193,6 +209,238 @@ rcl_subscription_get_default_options()
+}
+
+rcl_ret_t
+rcl_subscription_options_set_content_filter_options(
```
note to self: continue reviewing from here

ivanpauno reviewed (commented):
partial review

iuhilnehc-ynos commented on rcl/include/rcl/subscription.h:
```diff
@@ -53,6 +53,15 @@ typedef struct rcl_subscription_options_s
+{
+  /// Custom allocator for the options, used for incidental allocations.
+  /** For default behavior (malloc/free), see: rcl_get_default_allocator() */
+  rcl_allocator_t allocator;
```
No, because the following two functions https://github.com/ros2/rcl/blob/cd8e85931a92f3bdb25fb879c6df43e03cfbba75/rcl/include/rcl/subscription.h#L395-L397 and  https://github.com/ros2/rcl/blob/cd8e85931a92f3bdb25fb879c6df43e03cfbba75/rcl/include/rcl/subscription.h#L426-L428 use this structure directly.

ivanpauno replied:
But you're still not using this allocator there

https://github.com/ros2/rcl/blob/cd8e85931a92f3bdb25fb879c6df43e03cfbba75/rcl/src/rcl/subscription.c#L396

iuhilnehc-ynos replied:
Yeah, this line could be removed

ivanpauno replied:
> Yeah, this line could be removed

Could you do that?
It seems that it's currently unused, and if we delete it we can make sure we're actually always using the same allocator (the one provided by `rcl_subscription_options_t`).

iuhilnehc-ynos replied:
What I said is to remove the line you mentioned at https://github.com/ros2/rcl/pull/894#discussion_r836522760, not to remove the `rcl_allocator_t allocator;` in the `rcl_subscription_content_filter_options_t`.

ivanpauno replied:
Why not?
It's not being used anywhere (and if it is, use the one in `rcl_subscription_options_t` instead)

iuhilnehc-ynos replied:
Updated at 05811f0e5d1cbc72cca2abb212b3f06fb7e57f71

Barry-Xu-2018 commented:
LGTM

fujitatomoya commented on rcl/include/rcl/subscription.h:
```diff
@@ -209,6 +218,216 @@ RCL_WARN_UNUSED
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_subscription_options_fini(rcl_subscription_options_t * option);
```
@ivanpauno i will do that. i left comment on https://github.com/ros2/design/pull/282#issue-610568380, not to miss it for Humble Release Note.

fujitatomoya commented on rcl/test/rcl/test_subscription_content_filter_options.cpp:
```diff
@@ -0,0 +1,268 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
```
nit

```suggestion
// Copyright 2022 Open Source Robotics Foundation, Inc.
```

iuhilnehc-ynos replied:
updated in 0fcdc5b9a58a4e2f15dc3784abb8ff81e2f2f099

fujitatomoya commented on rcl/test/rcl/test_subscription.cpp:
```diff
@@ -885,6 +886,431 @@ TEST_F(CLASSNAME(TestSubscriptionFixture, RMW_IMPLEMENTATION), test_bad_return_l
     rcl_subscription_fini(&subscription, this->node_ptr)) << rcl_get_error_string().str;
 }
 
+/* A subscription with a content filtered topic setting.
```
mostly filtering test is checked with string type only, i think it is okay for primitive type, but we would want to test with user defined message just in case?

iuhilnehc-ynos replied:
The `BasicTypes.msg` is a user-defined message in `test_interface_files`, do you mean we need to test with the `Nested.msg`?

iuhilnehc-ynos replied:
Ah, I used the string in the test, I misremembered the BasicTypes is tested in `test_rmw_implementation`.
I will update the following case(`test_subscription_not_initialized_with_content_filtering`) by using the BasicTypes.

iuhilnehc-ynos replied:
Updated in 206facde4e4621d2af045e3ff176758f81ad6fda

fujitatomoya reviewed (approved):
lgtm with questions for further enhancement.

fujitatomoya commented:
> LGTM

@Barry-Xu-2018 could you give it a approval as community review, if it is LGTM to you?

Barry-Xu-2018 commented:
@fujitatomoya 

> > LGTM
> 
> @Barry-Xu-2018 could you give it a approval as community review, if it is LGTM to you?

I not find `approval` button.  
Maybe it is because I am not listed in reviewers.

ivanpauno commented on rcl/src/rcl/subscription.c:
```diff
@@ -193,6 +209,238 @@ rcl_subscription_get_default_options()
+      return rcl_convert_rmw_ret_to_rcl_ret(rmw_ret);
+    }
+
+    allocator->deallocate(
+      options->rmw_subscription_options.content_filter_options, allocator->state);
```
could we just set the content filter options in the previously allocated storage instead of allocating a new one and deallocating the old?

iuhilnehc-ynos replied:
This can avoid updating the original options while some failure happened in the previous.

ivanpauno replied:
You can still ensure that by copying the original options to a variable in the stack using https://github.com/ros2/rmw/blob/01d0db60c5bac242fe6f8a6d49494512f1e20c61/rmw/include/rmw/subscription_content_filter_options.h#L106, then set the new filter options, and if that fails you copy again the previous backup.

iuhilnehc-ynos replied:
Using your method can copy back the original content, but the pointer is changed.

ivanpauno replied:
> Using your method can copy back the original content, but the pointer is changed.

What pointer?

iuhilnehc-ynos replied:
Update at  d421cf267061020cd3e3bd8a519f75538f62a415.

ivanpauno commented on rcl/src/rcl/subscription.c:
```diff
@@ -193,6 +209,238 @@ rcl_subscription_get_default_options()
+rcl_subscription_content_filter_options_t
+rcl_get_zero_initialized_subscription_content_filter_options()
+{
+  return (const rcl_subscription_content_filter_options_t) {
+           .allocator = rcl_get_default_allocator(),
+           .rmw_subscription_content_filter_options =
+             rmw_get_zero_initialized_content_filter_options()
+  };  // NOLINT(readability/braces): false positive
```
indentantion looks weird here

```suggestion
  return (const rcl_subscription_content_filter_options_t) {
      .allocator = rcl_get_default_allocator(),
      .rmw_subscription_content_filter_options =
        rmw_get_zero_initialized_content_filter_options()
  };  // NOLINT(readability/braces): false positive
```

iuhilnehc-ynos replied:
The checking will be failed in `ament_uncrustify`.
```
@@ -301,3 +301,3 @@
-      .allocator = rcl_get_default_allocator(),
-      .rmw_subscription_content_filter_options =
-        rmw_get_zero_initialized_content_filter_options()
+           .allocator = rcl_get_default_allocator(),
+           .rmw_subscription_content_filter_options =
+             rmw_get_zero_initialized_content_filter_options()
```

ivanpauno commented on rcl/src/rcl/subscription.c:
```diff
@@ -193,6 +209,238 @@ rcl_subscription_get_default_options()
+    return RCL_RET_SUBSCRIPTION_INVALID;
+  }
+
+  rcl_allocator_t * allocator = (rcl_allocator_t *)&subscription->impl->options.allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
```
`allocator` is not being used in this function at all, so checking it doesn't seem to make sense

```suggestion
  rcl_allocator_t * allocator = (rcl_allocator_t *)&subscription->impl->options.allocator;
  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
```

iuhilnehc-ynos replied:
Yes, this will be removed as mentioned [before](https://github.com/ros2/rcl/pull/894#discussion_r836967104).

ivanpauno commented on rcl/src/rcl/subscription_impl.h:
```diff
@@ -19,11 +19,15 @@
 
 #include "rcl/subscription.h"
 
+/// Common content filter. (Defined in other package)
+typedef struct common_content_filter_s common_content_filter_t;
+
 struct rcl_subscription_impl_s
 {
   rcl_subscription_options_t options;
   rmw_qos_profile_t actual_qos;
   rmw_subscription_t * rmw_handle;
+  common_content_filter_t * common_content_filter;
```
is this needed now?

iuhilnehc-ynos replied:
OK, due to not breaking ABI, I'll remove the and `common_content_filter_factory_t` and  `common_content_filter_t`.

ivanpauno reviewed (commented):
part 2, still need to take a third look

iuhilnehc-ynos commented on rcl/src/rcl/subscription.c:
```diff
@@ -193,6 +209,238 @@ rcl_subscription_get_default_options()
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+
+  rcl_ret_t ret;
+  rmw_subscription_content_filter_options_t * content_filter_options =
+    allocator->allocate(
+    sizeof(rmw_subscription_content_filter_options_t), allocator->state);
```
The checking will be failed in `ament_uncrustify`.
```
      <failure message="Diff with 12 lines"><![CDATA[--- src/rcl/subscription.c
+++ src/rcl/subscription.c.uncrustify
@@ -253 +253 @@
-      sizeof(rmw_subscription_content_filter_options_t), allocator->state);
+    sizeof(rmw_subscription_content_filter_options_t), allocator->state);
```

ivanpauno replied:
that's weird, the current indentation sounds good then

ivanpauno commented on rcl/test/rcl/test_subscription.cpp:
```diff
@@ -885,6 +886,431 @@ TEST_F(CLASSNAME(TestSubscriptionFixture, RMW_IMPLEMENTATION), test_bad_return_l
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  }
+
+  if (is_cft_support) {
+    ASSERT_FALSE(wait_for_subscription_to_be_ready(&subscription, context_ptr, 30, 100));
+  } else {
+    ASSERT_TRUE(wait_for_subscription_to_be_ready(&subscription, context_ptr, 30, 100));
+
+    test_msgs__msg__Strings msg;
+    test_msgs__msg__Strings__init(&msg);
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+    {
+      test_msgs__msg__Strings__fini(&msg);
+    });
+    ret = rcl_take(&subscription, &msg, nullptr, nullptr);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ASSERT_EQ(
+      std::string(test_string),
+      std::string(msg.string_value.data, msg.string_value.size));
```
shouldn't we return early if cft is not supported?

Lines 944 onwards doesn't seem to do anything useful if cft is not supported.

iuhilnehc-ynos replied:
It's for testing the case(If the cft is not supported, it could definitely get the event and take the data.) even if the cft is not supported.

iuhilnehc-ynos replied:
It continue to test the unsupported case for set_content_filter at https://github.com/ros2/rcl/blob/0fcdc5b9a58a4e2f15dc3784abb8ff81e2f2f099/rcl/test/rcl/test_subscription.cpp#L1005

ivanpauno commented on rcl/test/rcl/test_subscription.cpp:
```diff
@@ -885,6 +886,431 @@ TEST_F(CLASSNAME(TestSubscriptionFixture, RMW_IMPLEMENTATION), test_bad_return_l
+ */
+TEST_F(
+  CLASSNAME(
+    TestSubscriptionFixture, RMW_IMPLEMENTATION), test_subscription_not_content_filtered_at_begin) {
```
```suggestion
    TestSubscriptionFixture, RMW_IMPLEMENTATION), test_subscription_not_initialized_with_content_filtering) {
```

iuhilnehc-ynos replied:
Thanks, updated in 206facde4e4621d2af045e3ff176758f81ad6fda

ivanpauno commented on rcl/test/rcl/test_subscription.cpp:
```diff
@@ -885,6 +886,431 @@ TEST_F(CLASSNAME(TestSubscriptionFixture, RMW_IMPLEMENTATION), test_bad_return_l
+    ret = rcl_subscription_set_content_filter(
+      &subscription, &options);
+    if (RCL_RET_UNSUPPORTED == ret) {
+      is_cft_support = false;
```
currently this fails with fastrtps but it's successful with connextdds, right?

iuhilnehc-ynos replied:
No, both rmw_fastrtps_cpp and rmw_connextdds are supported.

ivanpauno replied:
could we use the compile definition `RMW_IMPLEMENTATION` to make sure this doesn't fail for fastrtps/connextdds

iuhilnehc-ynos replied:
The test doesn't fail. What I mean `fails with rmw_cyclonedds_cpp` is RCL_RET_UNSUPPORTED. Using `colcon test` should not get a failure.

ivanpauno replied:
Yes.
What I mean is to have an `ASSERT_EQ(ret, RCL_RET_OK)` for both fastrtps and connextdds.

ivanpauno replied:
I see you have already done that in https://github.com/ros2/rcl/pull/894/commits/630113369c87433e0bfe5d0f18b8ec73c267f7c7.

ivanpauno commented on rcl/include/rcl/subscription.h:
```diff
@@ -324,13 +324,15 @@ rcl_subscription_content_filter_options_init(
 RCL_WARN_UNUSED
 rcl_ret_t
 rcl_subscription_content_filter_options_set(
+  const rcl_subscription_t * subscription,
```
can this be `rcl_allocator_t`?

iuhilnehc-ynos replied:
[rcl_subscription_content_filter_options_init](https://github.com/ros2/rcl/blob/05811f0e5d1cbc72cca2abb212b3f06fb7e57f71/rcl/include/rcl/subscription.h#L301) the `rcl_subscription_content_filter_options_t` with the rcl_subscription by using the allocator in the subscription_options,  but to use a specific `rcl_allocator_t` for [rcl_subscription_content_filter_options_set](https://github.com/ros2/rcl/blob/05811f0e5d1cbc72cca2abb212b3f06fb7e57f71/rcl/include/rcl/subscription.h#L334) and [rcl_subscription_content_filter_options_fini](https://github.com/ros2/rcl/blob/05811f0e5d1cbc72cca2abb212b3f06fb7e57f71/rcl/include/rcl/subscription.h#L334)?  The allocator in the rcl_subscriptions is not available for users. I think it's not convinent for user to call these two functions with a `rcl_allocator_t`. (e.x. how to fini the rcl_subscription_content_filter_options_t [here](https://github.com/ros2/rcl/blob/05811f0e5d1cbc72cca2abb212b3f06fb7e57f71/rcl/test/rcl/test_subscription.cpp#L1011-L1012)?)

ivanpauno replied:
> [rcl_subscription_content_filter_options_init](https://github.com/ros2/rcl/blob/05811f0e5d1cbc72cca2abb212b3f06fb7e57f71/rcl/include/rcl/subscription.h#L301) the rcl_subscription_content_filter_options_t with the rcl_subscription by using the allocator in the subscription_options, but to use a specific rcl_allocator_t for [rcl_subscription_content_filter_options_set](https://github.com/ros2/rcl/blob/05811f0e5d1cbc72cca2abb212b3f06fb7e57f71/rcl/include/rcl/subscription.h#L334) and [rcl_subscription_content_filter_options_fini](https://github.com/ros2/rcl/blob/05811f0e5d1cbc72cca2abb212b3f06fb7e57f71/rcl/include/rcl/subscription.h#L334)?

I meant to change it in the three.
Though your argument sounds good.

ivanpauno commented on rcl/include/rcl/subscription.h:
```diff
@@ -346,15 +348,18 @@ rcl_subscription_content_filter_options_set(
 RCL_WARN_UNUSED
 rcl_ret_t
 rcl_subscription_content_filter_options_fini(
+  const rcl_subscription_t * subscription,
```
same

fujitatomoya commented:
CI seems to be unstable, retry just once again.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=16493)](http://ci.ros2.org/job/ci_linux/16493/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=11081)](http://ci.ros2.org/job/ci_linux-aarch64/11081/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=16866)](http://ci.ros2.org/job/ci_windows/16866/)

fujitatomoya commented:
@iuhilnehc-ynos CI is unstable, do you have the same failure in your local env?

iuhilnehc-ynos commented:
@fujitatomoya 

> do you have the same failure in your local env?

Sorry, I didn't run the test with `--packages-above rcl rclcpp` before.

I'll check if the following [issues](https://ci.ros2.org/job/ci_linux/16486/#showFailuresLink) could be reproduced in my local env or not,
[projectroot.test_showimage_cam2image__rmw_connextdds](https://ci.ros2.org/job/ci_linux/16486/testReport/junit/(root)/projectroot/test_showimage_cam2image__rmw_connextdds/)
[projectroot.test_showimage_cam2image__rmw_cyclonedds_cpp](https://ci.ros2.org/job/ci_linux/16486/testReport/junit/(root)/projectroot/test_showimage_cam2image__rmw_cyclonedds_cpp/)
[projectroot.test_showimage_cam2image__rmw_fastrtps_cpp](https://ci.ros2.org/job/ci_linux/16486/testReport/junit/(root)/projectroot/test_showimage_cam2image__rmw_fastrtps_cpp/)

iuhilnehc-ynos commented:
@fujitatomoya 

I can reproduce these issues in my local env because the macro `CONTENT_FILTER_ENABLE` in [rclcpp](https://github.com/ros2/rclcpp/pull/1561) doesn't work. Updated in https://github.com/ros2/rclcpp/pull/1561/commits/1b232aeedc96d065315b95bb902ca86ed4b8357e

iuhilnehc-ynos commented:
rerun CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=16499)](http://ci.ros2.org/job/ci_linux/16499/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=11088)](http://ci.ros2.org/job/ci_linux-aarch64/11088/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=16872)](http://ci.ros2.org/job/ci_windows/16872/)

fujitatomoya commented:
@iuhilnehc-ynos CI looks much better 👍 but still some failure(missing result) observed.

ivanpauno commented on rcl/src/rcl/subscription.c:
```diff
@@ -193,6 +209,290 @@ rcl_subscription_get_default_options()
+  rcl_subscription_options_t * options)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(filter_expression, RCL_RET_INVALID_ARGUMENT);
+  if (expression_parameters_argc > 100) {
+    RCL_SET_ERROR_MSG("The maximum of expression parameters argument number is 100");
+    return RCL_RET_INVALID_ARGUMENT;
+  }
```
isn't this check already done in rmw?

iuhilnehc-ynos replied:
Considering rmw_connextdds has its [special extension setting](https://github.com/ros2/rmw/pull/302#discussion_r827572015) and a fallback for cft in the `rcl`, so I need to add this checkment in `rcl`.

ivanpauno commented on rcl/src/rcl/subscription.c:
```diff
@@ -193,6 +209,290 @@ rcl_subscription_get_default_options()
+  RCL_CHECK_ARGUMENT_FOR_NULL(options, RCL_RET_INVALID_ARGUMENT);
+  const rcl_allocator_t * allocator = &subscription->impl->options.allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+  if (expression_parameters_argc > 100) {
+    RCL_SET_ERROR_MSG("The maximum of expression parameters argument number is 100");
+    return RCL_RET_INVALID_ARGUMENT;
+  }
```
same

ivanpauno commented on rcl/src/rcl/subscription.c:
```diff
@@ -193,6 +209,290 @@ rcl_subscription_get_default_options()
+  if (!rcl_subscription_is_valid(subscription)) {
+    return RCL_RET_SUBSCRIPTION_INVALID;
+  }
+  if (expression_parameters_argc > 100) {
+    RCL_SET_ERROR_MSG("The maximum of expression parameters argument number is 100");
+    return RCL_RET_INVALID_ARGUMENT;
+  }
```
same

ivanpauno reviewed (approved):
LGTM with green CI

iuhilnehc-ynos commented:
The issue about https://ci.ros2.org/job/ci_linux/16499/testReport/(root)/projectroot/test_play_services__rmw_fastrtps_cpp/ can't be reproduced in my local env. 

[Wait for a service](https://github.com/ros2/rosbag2/blob/e156eb9e38c377df33e5075ab7fa48f3b52eaa12/rosbag2_transport/test/rosbag2_transport/test_play_services.cpp#L101) with [1s](https://github.com/ros2/rosbag2/blob/e156eb9e38c377df33e5075ab7fa48f3b52eaa12/rosbag2_transport/test/rosbag2_transport/test_play_services.cpp#L205) seems a bit smaller to wait a service if [calling services](https://github.com/ros2/rosbag2/blob/e156eb9e38c377df33e5075ab7fa48f3b52eaa12/rosbag2_transport/test/rosbag2_transport/test_play_services.cpp#L329) so frequently.

I have addressed issues in Linux-aarch64 and Windows.

Give it another shot.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=16509)](http://ci.ros2.org/job/ci_linux/16509/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=11096)](http://ci.ros2.org/job/ci_linux-aarch64/11096/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=16880)](http://ci.ros2.org/job/ci_windows/16880/)

iuhilnehc-ynos commented:
It seems that waiting for some events ([matched a pair for sub and pub](https://github.com/ros2/rosbag2/blob/e156eb9e38c377df33e5075ab7fa48f3b52eaa12/rosbag2_transport/test/rosbag2_transport/test_record.cpp#L107) in [1](https://ci.ros2.org/job/ci_windows/16880/testReport/junit/rosbag2_transport/RecordIntegrationTestFixture/qos_is_stored_in_metadata/), [wait for a message event](https://github.com/iuhilnehc-ynos/rcl/blob/954982ee240cdc305302bd3163bcab3ebcf8933a/rcl/test/rcl/test_subscription.cpp#L1300) in [2](https://ci.ros2.org/job/ci_windows/16880/testReport/junit/rcl/TestSubscriptionFixture__rmw_connextdds/test_subscription_not_initialized_with_content_filtering/), etc) with 10s[[1](https://github.com/ros2/rosbag2/blob/e156eb9e38c377df33e5075ab7fa48f3b52eaa12/rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp#L109)] is not enough for rmw_connextdds in Windows, I don't know if I should increase the timeout more (which might/might not work) or not ?

I think that solving such an issue needs help from the `connextdds/rmw_connextdds` group.

iuhilnehc-ynos commented:
@asorbini 

About https://github.com/ros2/rcl/pull/894#issuecomment-1086573960, do you have any suggestions?

iuhilnehc-ynos commented:
unstable tests in windows are unrelated to current PR, merge it.

Blast545 commented:
:man_farmer: I think there are is least 1 new test regressions shown in the windows job related to this PR:

https://ci.ros2.org/job/ci_windows/16880/testReport/junit/rcl/TestSubscriptionFixture__rmw_connextdds/test_subscription_not_initialized_with_content_filtering/

And now in the buildfarm:
https://ci.ros2.org/view/nightly/job/nightly_win_rel/2261/testReport/junit/rcl/TestSubscriptionFixture__rmw_connextdds/test_subscription_not_initialized_with_content_filtering/
